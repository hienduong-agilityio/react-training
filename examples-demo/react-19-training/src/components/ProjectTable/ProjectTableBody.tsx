// Component

// Type
import type { IProjectItemProps } from '../../interfaces';
import ProjectItem from '../ProjectItem';

export interface IProjectTableBodyProps {
  // tableData: The data for each project to be rendered.
  tableData: IProjectItemProps[];
  // onDeleteProject: The function to open delete modal
  onDeleteProject: (projectId: string) => void;
}

/**
 * ProjectTableBody Component
 *
 * @returns {JSX.Element} The table body containing the project items or a "No results found" message.
 */
const ProjectTableBody = ({ tableData = [], onDeleteProject = () => {} }: IProjectTableBodyProps) => {
  const flattenedData = tableData.map((project) => {
    const { id, projectName, status, manager, timeline, lastUpdate, resources, budget } = project;

    return {
      id: id,
      projectName: projectName,
      status: status,
      managerName: manager?.managerName,
      managerImage: manager?.managerImage,
      lastUpdate: lastUpdate,
      resources: resources,
      timeStart: timeline?.timeStart,
      timeEnd: timeline?.timeEnd,
      budget: budget
    };
  });

  return (
    <tbody>
      {tableData.length > 0 ? (
        flattenedData.map((project: IProjectItemProps) => {
          return <ProjectItem key={project.id} {...project} onDeleteProject={onDeleteProject} />;
        })
      ) : (
        <tr>
          <td colSpan={8} className='text-center py-4'>
            No results found
          </td>
        </tr>
      )}
    </tbody>
  );
};

export default ProjectTableBody;
