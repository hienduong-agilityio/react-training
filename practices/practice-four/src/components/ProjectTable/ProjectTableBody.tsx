// Component
import ProjectItem from '@/components/ProjectItem';

// Type
import type { IProjectItemProps } from '@/interfaces';
import { Spinner } from '@/components';

export interface IProjectTableBodyProps {
  // tableData: The data for each project to be rendered.
  tableData: IProjectItemProps[];
  // onDeleteProject: The function to open delete modal
  onDeleteProject: (projectId: string) => void;
  isPending?: boolean;
}

/**
 * ProjectTableBody Component
 *
 * @returns {JSX.Element} The table body containing the project items or a "No results found" message.
 */
const ProjectTableBody = ({ tableData = [], onDeleteProject = () => {}, isPending }: IProjectTableBodyProps) => {
  const flattenedData = tableData.map((project) => {
    const { id, projectName, status, manager, timeline, lastUpdate, resources, budget } = project;

    return {
      id: id,
      projectName: projectName,
      status: status,
      managerName: manager?.managerName,
      managerImage: manager?.managerImage,
      lastUpdate: lastUpdate,
      resources: resources,
      timeStart: timeline?.timeStart,
      timeEnd: timeline?.timeEnd,
      budget: budget
    };
  });

  return (
    <tbody className={isPending ? 'pointer-events-none select-none relative' : ''}>
      {isPending && (
        <div className='absolute inset-0 z-10 flex justify-center items-center bg-white/20 backdrop-blur-[1px] pointer-events-auto'>
          <Spinner />
        </div>
      )}
      {tableData.length > 0 ? (
        flattenedData.map((project: IProjectItemProps) => {
          return <ProjectItem key={project.id} {...project} onDeleteProject={onDeleteProject} />;
        })
      ) : (
        <tr>
          <td colSpan={8} className='text-center py-4'>
            No results found
          </td>
        </tr>
      )}
    </tbody>
  );
};

export default ProjectTableBody;
