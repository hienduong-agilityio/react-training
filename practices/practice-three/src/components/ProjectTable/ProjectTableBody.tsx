// Component
import ProjectItem from '@/components/ProjectItem';

// Type
import type { IProjectItemProps } from '@/components/ProjectItem';

export interface IProjectTableBodyProps {
  // tableData: The data for each project to be rendered.
  tableData: IProjectItemProps[];
  // onOpenEdit: The function to open edit modal
  onOpenEdit: (id: string) => void;
}

/**
 * ProjectTableBody Component
 *
 * @returns {JSX.Element} The table body containing the project items.
 */
const ProjectTableBody = ({ tableData, onOpenEdit }: IProjectTableBodyProps) => {
  return (
    <tbody>
      {tableData.map((project: IProjectItemProps, index: number) => (
        <ProjectItem key={project.id} {...project} index={index + 1} onOpenEdit={onOpenEdit} />
      ))}
    </tbody>
  );
};

export default ProjectTableBody;
