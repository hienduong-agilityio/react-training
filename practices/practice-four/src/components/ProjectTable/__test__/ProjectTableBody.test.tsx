// Libraries
import { render, screen, RenderResult } from '@testing-library/react';
import ProjectTableBody from '@/components/ProjectTable/ProjectTableBody';
import { MemoryRouter } from 'react-router-dom';

// Interfaces
import type { IProjectItemProps } from '@/interfaces';

const mockTableData: IProjectItemProps[] = [
  {
    id: '123',
    projectName: 'Test Project 1',
    status: 'In Progress',
    manager: {
      managerName: 'John Doe',
      managerImage: '/path/to/image.jpg'
    },
    timeline: {
      timeStart: '2024-01-01',
      timeEnd: '2024-12-31'
    },
    lastUpdate: '2024-10-10',
    resources: '5',
    budget: 10000
  }
];

// Mock onDeleteProject function
const mockOnDeleteProject = jest.fn();

describe('ProjectTableBody Component', () => {
  let renderComponent: (tableData?: IProjectItemProps[]) => RenderResult;

  beforeEach(() => {
    renderComponent = (tableData = mockTableData) => {
      return render(
        <MemoryRouter>
          <ProjectTableBody tableData={tableData} onDeleteProject={mockOnDeleteProject} />
        </MemoryRouter>
      );
    };
  });

  it('matches the snapshot when no data is provided', () => {
    const { container } = renderComponent([]);
    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot when data is provided', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('renders project items when tableData is provided', () => {
    renderComponent();

    expect(screen.getByText('Test Project 1')).toBeInTheDocument();
    expect(screen.getByAltText('John Doe avatar')).toBeInTheDocument();
  });

  it('renders "No results found" when tableData is empty', () => {
    renderComponent([]);

    expect(screen.getByText('No results found')).toBeInTheDocument();
  });
});
