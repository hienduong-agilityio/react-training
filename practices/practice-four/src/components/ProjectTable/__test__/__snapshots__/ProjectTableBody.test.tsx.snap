// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectTableBody Component matches the snapshot when data is provided 1`] = `
<div>
  <tbody>
    <tr
      class="border-b border-gray-200 hover:bg-light group"
    >
      <td
        class="px-3 py-5 font-medium"
      >
        <p>
          123
        </p>
      </td>
      <td>
        <a
          class="text-primary-500 font-semibold hover:underline cursor-pointer w-max"
          href="/projects/123"
        >
          Test Project 1
        </a>
      </td>
      <td
        class="text-center px-2"
      >
        <div
          class="inline-flex border-2 border-gray-100 items-center justify-center overflow-hidden bg-primary-0 rounded-lg w-7 h-7"
        >
          <img
            alt="John Doe avatar"
            class="h-full w-full object-cover"
            src="/path/to/image.jpg"
          />
        </div>
      </td>
      <td>
        <div
          class="w-max"
        >
          <span
            class="text-xs py-[1px] px-2 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
          >
            <svg
              class="inline-flex color-default"
              fill="none"
              height="6"
              role="img"
              viewBox="0 0 6 6"
              width="6"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="6"
                rx="2"
                width="6"
              />
            </svg>
            In Progress
          </span>
        </div>
      </td>
      <td>
        <p
          class="text-gray-600 w-max"
        >
          2024-10-10
        </p>
      </td>
      <td
        class="text-center px-2"
      >
        <div>
          <span
            class="py-[3px] px-2 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
          >
            5
          </span>
        </div>
      </td>
      <td>
        <div
          class="flex gap-1 items-center"
        >
          <span
            class="w-[98px] h-5 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
          >
            2024-01-01
          </span>
          <svg
            aria-label="Next"
            fill="none"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.5 11L9.5 8L6.5 5"
              stroke="#868FA0"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          <span
            class="w-[98px] h-5 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
          >
            2024-12-31
          </span>
        </div>
      </td>
      <td>
        <div
          class="flex items-center w-44 justify-between"
        >
          <p
            class="text-primary-400 font-bold"
          >
            US$ 10000k
          </p>
          <div
            class="group-hover:block hidden"
          >
            <div
              class="relative inline-block text-left group"
            >
              <button
                class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 border-white text-primary-400 hover:border-2 hover:border-primary-100 px-4 py-2 text-base text-black hover:border-none hover:bg-gray-100"
              >
                <img
                  alt="Menu"
                />
              </button>
            </div>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</div>
`;

exports[`ProjectTableBody Component matches the snapshot when no data is provided 1`] = `
<div>
  <tbody>
    <tr>
      <td
        class="text-center py-4"
        colspan="8"
      >
        No results found
      </td>
    </tr>
  </tbody>
</div>
`;
